### PR TITLE
colexec: extract multiple new packages

### PR DESCRIPTION
**Pull request by yuzefovich**
This PR breaks down huge `colexec` package in order to speed up the build time
(it was pointed out that the package is the bottleneck during bazel builds). With
all of these changes, `colexec` (and its descendants) is barely visible when profiling
the bazel build.

The structure of dependencies is enforced by `dep_test` files added to the new
packages.

The following new packages have been extracted:
- `colexecutils` which contains miscellaneous utility operators, structs,
and functions that will be used by several other packages
- `colexechash` which contains all of the code interacting directly
with hashing of data (the hash table and the tuple hash distributor)
- `colexecwindow` which contains all of the code related to window
functions
- `colexecargs` which contains the arguments to and the result of
`NewColOperator` call
- `colexecbase` which contains miscellaneous operators that have very
few dependencies and are depended on by other operators that we want
to extract from `colexec`
- `colexeccmp` which contains shared objects between projection and
selection operators (namely, LIKE ops- and default comparison- related
things)
- `colexecproj` and `colexecsel` which contain the projection and the
selection operators, respectively
- `colexecjoin` which contains the code for the in-memory joiners (cross,
hash, and merge).

Additionally, multiple other operators have been moved to more appropriate
packages as well as the following package moves and renames were performed:
- move `colexecerror` out of `sql/colexecbase` into `sql`
- rename `sql/colexecbase` to `sql/colexecop`.

See individual commits for details.

### CODE REVIEW

**Comment by yuzefovich** [comment]
I think this is RFAL. The first commit should be ignored since it is a part of a separate PR.

File: pkg/sql/colexec/colexecmisc/COLEXECMISC.bzl
Comment (rickystewart) [inline pkg/sql/colexec/colexecmisc/COLEXECMISC.bzl:1]: I think this is too much code duplication. The helpers should be refactored into a common place. Let me know if you need help with this.

**Review by rickystewart — DISMISSED** [review]

**Review by irfansharif — COMMENTED** [review]
If there's already a common ("base") package that all these other packages depend on, the common bazel rules can be moved in there, and that would avoid the deduplication. If there isn't, we could create one (even if only for bazel). Ultimately the "bazel stuff" within colexec will be/is already owned by the sql exec team, so I'll defer to y'all for what you think is best. 

In any case, thanks for chopping all of this up, it's pretty great! Very excited to try it out.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60770#-:-MUAAR0z24TtsUp9Q1nt:b-49t729)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

**Comment by asubiotto** [comment]
> colexecmisc which contains miscellaneous operators that have very
few dependencies and are depended on by other operators that we want
to extract from colexec

can we put the colexecmisc stuff into colexecbase? Seems like both packages serve the same purpose.

> colexecprojsel which contains shared objects between projection and
selection operators (namely, LIKE ops- and default comparison- related
things)

😬  Don't think it's a good idea to have `colexecproj`, `colexecsel`, and `colexecprojsel`. Is there something better we can name it? Maybe `colexeccmp`?

**Comment by yuzefovich** [comment]
> can we put the `colexecmisc` stuff into `colexecbase`? Seems like both packages serve the same purpose.

No, they serve different purposes. `colexecbase` defines the `Operator` interface whereas `colexecmisc` contains different things implementing that interface. Merging those two would be bad since it would introduce the dependency on the new stuff that `colexecmisc` brings into the users of `colexecbase` (which are pretty much all packages).

> Don't think it's a good idea to have `colexecproj`, `colexecsel`, and `colexecprojsel`. Is there something better we can name it? Maybe `colexeccmp`?

I disagree because both projection and selection operators result in huge amount of generated code, and they become the bottleneck. In fact, we might want to go even further - break down `colexecproj` into `colexecprojconst` and `colexecprojnonconst` or something like that, so that we can separate out huge projection_ops files (but this can wait).

**Comment by asubiotto** [comment]
> No, they serve different purposes. colexecbase defines the Operator interface whereas colexecmisc contains different things implementing that interface. Merging those two would be bad since it would introduce the dependency on the new stuff that colexecmisc brings into the users of colexecbase (which are pretty much all packages).

Well then let's rename `colexecbase` to be `colexecop` and `colexecmisc` to be `colexecbase`. `misc` feels too vague.

> I disagree because both projection and selection operators result in huge amount of generated code, and they become the bottleneck. In fact, we might want to go even further - break down colexecproj into colexecprojconst and colexecprojnonconst or something like that, so that we can separate out huge projection_ops files (but this can wait).

I'm not talking about the number of packages. I think that's good. I specifically mean changing the `colexecprojsel` name.

**Comment by yuzefovich** [comment]
Renaming `colexecbase` to `colexecop` sounds good to me. I think I will also extract `colexecerror` outside of `sql/colexecbase` folder and put it in `sql` folder.

I misunderstood your comment about `colexecprojsel`, renaming it sounds good.

**Comment by yuzefovich** [comment]
First commit is #60993.

I removed the duplication of bazel-related things and renamed `colexecprojsel` to `colexeccmp`.

However, I'm encountering some issues when renaming/moving packages with `make bazel-generate` (first I tried to move `colexecerror` out of `sql/colexecbase` into `sql`, and then independently I tried renaming `colexecbase` to `colexecop`, and both times I'm getting some warnings locally, but the bazel build succeeds). I don't want to just rename `colexecmisc` to `colexecbase` because then we will have package name clash between `sql/colexecbase` and `sql/colexec/colexecbase`, so I think we should proceed with this PR as is and perform the renaming later.

**Comment by asubiotto** [comment]
OK. To be clear, why will it be easier later? In any case, LGTM. We should probably merge this sooner rather than later to avoid having to rebase too much.

**Comment by yuzefovich** [comment]
I'm not saying it will be easier to do the rename later, but that we should not wait for it to happen given that rebasing this PR is a pain. I don't understand why I'm seeing those warnings locally, so I'll need help from dev-infra team.

**Comment by irfansharif** [comment]
> However, I'm encountering some issues when renaming/moving packages with make bazel-generate (first I tried to move colexecerror out of sql/colexecbase into sql, and then independently I tried renaming colexecbase to colexecop, and both times I'm getting some warnings locally, but the bazel build succeeds)

What are the warnings? It'd be helpful if you could show us how to repro it (is it a particular commit + some particular command)? I'm just going to rubber stamp this PR on Alfonso's behalf, reading his last comment.

**Review by irfansharif — APPROVED** [review]

**Comment by yuzefovich** [comment]
Ok, I figured out the problem with warnings (as it turned out I had old path being imported in a couple of test files, and Goland for some reason couldn't find them when I was doing the search - probably busted caches). Anyway, I added 3 more commits to get to the currently desired state:
1. move `colexecerror` out of `sql/colexecbase` into `sql`
2. rename `sql/colexecbase` to `sql/colexecop`
3. rename `colexec/colexecmisc` to `colexec/colexecbase`.

I'll wait for #60993 to merge and will merge this afterwards.

Thanks for the feedback everyone!

**Comment by yuzefovich** [comment]
Oops, forgot the label.

bors r+

**Pull request merged**
